channeldb/payment_control.go

```diff
@@ -223,6 +223,18 @@ func (p *PaymentControl) InitPayment(paymentHash lntypes.Hash,
 	return updateErr
 }
 
+// DeleteFailedAttempts deletes all failed htlcs for a payment.
+func (p *PaymentControl) DeleteFailedAttempts(hash lntypes.Hash) error {
+	if !p.db.keepFailedPaymentAttempts {
```

Comment:
Non-blocking, just something to think about. I think originally we have the flag `keepFailedPaymentAttempts` placed in `channeldb` rather than in `routing` because we want to introduce an optional migration that deletes the attempts, which makes total sense as it's db's responsibility to manage the migrations. On the other hand, it's nice to have "pure" CRUD methods in `channeldb` and have constraints made at the callsite to maximally generalize those CRUD methods. In our case, `DeleteFailedAttempts` can just do its job without checking the flag, and let `routing` decide whether to prune the HTLCs based on its own config, which means we'd have another flag placed in `routing` for that purpose. A big however is, since we already have the flag in db, it can perfectly serve the two purposes so we'd rather only have one flag there. I'll continue the onion-removal work on payments and invoices to see how it fit the whole picture. Atm, I think your approach is the best.

Reply:
I'm a bit unclear on the plan going forward. Are you saying we could move the flag from `channeldb` to `routing` after the optional migration is added?